### DISCUSSION TITLE
Meeting minutes for 22.08.2023

### DISCUSSION THREAD

**Discussion by konradweiss** [Fraunhofer-AISEC/cpg#1661] [Dev Meetings]
### Agenda

1. Open PRs
2. CPG Discussions open topics and next steps
3. CPG evaluation in Codyze executors and SARIF

### Discussed PRs

-[]

**Comment by oxisto** [comment]
I have the following discussion point with regard to https://github.com/Fraunhofer-AISEC/cpg/pull/1646.

The "bug" in neo4j is fixed (although they say it's completely unsupported) so we could directly use an AstEdge in a single-node property edge. However, this has the drawback that in order to support "on change" events, e.g. to replace a type listener when the edge changes, we need to resort back to overriding the setter. If we would stick with the singleton edge list, the list can take care of this like this:

```kotlin
var initializerEdge =
        astOptionalEdgeOf<Expression>(
            onChanged = { old, new ->
                val value = new?.end
                exchangeTypeObserver(old, new)
                if (value is Reference) {
                    value.resolutionHelper = this
                }
            }
        )
```

otherwise a setter would be needed like this:
```kotlin
var initializerEdge = astOptionalEdgeOf<Expression>()
set(new) {
                val value = new?.end
                exchangeTypeObserver(field, new)
                if (value is Reference) {
                    value.resolutionHelper = this
                }
               field = new
```

This only is needed in a couple of places (~ 10), but still not really pretty. The advantage of course is that we do not need a pointless wrapper around an `Edge`.

**Reply by konradweiss** [reply]
We keep the singleton edge lists as they better encapsulate the functionality and we dont have to overwrite the setters.

**Comment by konradweiss** [comment]
Singleton Edge List
- We keep the singleton edge lists as they better encapsulate the functionality and we don't have to overwrite the setters.

C/C++ Function resolution has issues
- signature based matching is used somewhere although the trait for overloading is not active in the language.
- Unknown matches apparently prevent matching because we find no function with a n argument type that is unknown. Unknown types should match to all types, because when the would be resolved to the correct type they may match.
Important feature extensions for the python frontend: syntax features fro lists and dicts.
